source/routing/next_stop_time.cpp

```diff
@@ -61,31 +61,42 @@ bool NextStopTimeData::Arrival::is_valid(const type::StopTime& st) const {
 
 template <typename Getter>
 void NextStopTimeData::TimesStopTimes<Getter>::init(const JourneyPattern& jp, const JourneyPatternPoint& jpp) {
+    using StopTimePair = std::pair<const type::StopTime*, const type::StopTime*>;
     // collect the stop times at the given jpp
     const auto jpp_order = jpp.order;
-    stop_times.reserve(jp.discrete_vjs.size());
+    std::vector<StopTimePair> stop_times_and_earliest_stop_time;
+    stop_times_and_earliest_stop_time.reserve(jp.discrete_vjs.size());
     for (const auto& vj : jp.discrete_vjs) {
-        const auto& st = get_corresponding_stop_time(*vj, jpp_order);
+        const type::StopTime& st = get_corresponding_stop_time(*vj, jpp_order);
         if (!getter.is_valid(st)) {
             continue;
         }
-        stop_times.push_back(&st);
+        const type::StopTime& earliest_stop_time = navitia::earliest_stop_time(st.vehicle_journey->stop_time_list);
+        stop_times_and_earliest_stop_time.emplace_back(&st, &earliest_stop_time);
     }
 
     // sort the stop times in ascending order
-    boost::sort(stop_times, [&](const type::StopTime* st1, const type::StopTime* st2) {
+    boost::sort(stop_times_and_earliest_stop_time, [&](StopTimePair pair1, StopTimePair pair2) {
+        const type::StopTime* st1 = pair1.first;
+        const type::StopTime* st2 = pair2.first;
         const auto time1 = DateTimeUtils::hour(getter.get_time(*st1));
         const auto time2 = DateTimeUtils::hour(getter.get_time(*st2));
         if (time1 != time2) {
             return time1 < time2;
         }
-        const auto& st1_first = navitia::earliest_stop_time(st1->vehicle_journey->stop_time_list);
-        const auto& st2_first = navitia::earliest_stop_time(st2->vehicle_journey->stop_time_list);
```

Comment:
How about infinite caching (hash_map) of this instead?
Theoretical complexity and principle would be the same, but it should save a lot of useless `earliest_stop_time()` calls search, no?
And also the final reallocation, but that's probably negligible.

Reply:
We could cache this, but I don't know where to put the cache. 
The cache would make sense only if it lives outside of the `init()` function : on a single call to `init()`, I don't expect to have many stop_times belonging to the same vj.

Reply:
Side-effects are not nice, but that's the only way I can think of for a shared cache here.

Another lead is to lower the cache and cache only the given stop-time, which would make the first comparison O(K) and let the other comparisons O(1) resulting in the same theoretical complexity than what you did, but saving all the cases where getting first-stop is useless, I think.

Reply:
Maybe onboard cache structure in NextStopTimeData, but clean it up at a smart time.

Reply:
> Another lead is to lower the cache and cache only the given stop-time, which would make the first comparison O(K) and let the other comparisons O(1) resulting in the same theoretical complexity than what you did, but saving all the cases where getting first-stop is useless, I think.

Yes, it may be slightly better, but I am not sure either: using a hashmap is slightly more expensive than a plain vector.
Anyway, I don't think the potential gains area worth the time spent trying.

However, I do think it is worth keeping in mind for future work the sorting of `vj.stop_time_list`, and removing the need to compute/cache the earliest stop_time